src/main/java/software/amazon/awssdk/crt/mqtt/MqttConnection.java

```diff
@@ -127,130 +115,129 @@ public ConnectionState getState() {
+            return future.thenApply(unused -> packetId);
         }
         catch (CrtRuntimeException ex) {
             throw new MqttException("AWS CRT exception: " + ex.toString());
         }
     }
 
-    public short unsubscribe(String topic) throws MqttException {
-        return unsubscribe(topic, null);
-    }
-
-    public short unsubscribe(String topic, MqttActionListener ack) throws MqttException {
+    public CompletableFuture<Integer> unsubscribe(String topic) throws MqttException {
         if (native_ptr() == 0) {
             throw new MqttException("Invalid connection during unsubscribe");
         }
 
-        AsyncCallback unsubAck = wrapAck(ack);
-        return mqtt_unsubscribe(native_ptr(), topic, unsubAck);
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        AsyncCallback unsubAck = wrapPacketFuture(future);
+        int packetId = mqttConnectionUnsubscribe(native_ptr(), topic, unsubAck);
+        // When the future completes, complete the returned future with the packetId
+        return future.thenApply(unused -> packetId);
     }
 
-    public short publish(MqttMessage message, QOS qos, boolean retain) throws MqttException {
-        return publish(message, qos, retain, null);
-    }
-
-    public short publish(MqttMessage message, QOS qos, boolean retain, MqttActionListener ack) throws MqttException {
+    public CompletableFuture<Integer> publish(MqttMessage message, QoS qos, boolean retain) throws MqttException {
         if (native_ptr() == 0) {
             throw new MqttException("Invalid connection during publish");
         }
-        
-        AsyncCallback pubAck = wrapAck(ack);
+
+        CompletableFuture<Integer> future = new CompletableFuture<>();
+        AsyncCallback pubAck = wrapPacketFuture(future);
         try {
-            return mqtt_publish(native_ptr(), message.getTopic(), qos.getValue(), retain, message.getPayloadDirect(), pubAck);
+            int packetId = mqttConnectionPublish(native_ptr(), message.getTopic(), qos.getValue(), retain, message.getPayloadDirect(), pubAck);
+            // When the future completes, complete the returned future with the packetId
+            return future.thenApply(unused -> packetId);
         }
         catch (CrtRuntimeException ex) {
             throw new MqttException("AWS CRT exception: " + ex.toString());
         }
     }
 
-    public boolean setWill(MqttMessage message, QOS qos, boolean retain) throws MqttException {
+    public boolean setWill(MqttMessage message, QoS qos, boolean retain) throws MqttException {
         if (native_ptr() == 0) {
             throw new MqttException("Invalid connection during setWill");
         }
 
         try {
-            return mqtt_set_will(native_ptr(), message.getTopic(), qos.getValue(), retain, message.getPayloadDirect());
+            return mqttConnectionSetWill(native_ptr(), message.getTopic(), qos.getValue(), retain, message.getPayloadDirect());
```

Comment:
this just returns True/False for success/failure. It should probably be void/throw

Reply:
ok.